File: src/features/animenight/animeNightRepository.js
const db = require('../../database/db');

// Inserts multiple anime entries in a single batch (all sharing the same watched
// date and "added at" timestamp, since they come from one /animenight add call).
async function addEntries(guildId, titles, watchedDate, addedBy) {
  await db.ready;
  const addedAt = Date.now();

  const statements = titles.map((title) => ({
    sql: `INSERT INTO anime_night_entries (guild_id, title, watched_date, added_at, added_by)
          VALUES (?, ?, ?, ?, ?)`,
    args: [guildId, title, watchedDate, addedAt, addedBy],
  }));

  await db.client.batch(statements, 'write');
}

async function getAllEntries(guildId) {
  await db.ready;
  const result = await db.client.execute({
    sql: 'SELECT * FROM anime_night_entries WHERE guild_id = ? ORDER BY id ASC',
    args: [guildId],
  });
  return result.rows;
}

async function getLastEntries(guildId, limit) {
  await db.ready;
  const result = await db.client.execute({
    sql: 'SELECT * FROM anime_night_entries WHERE guild_id = ? ORDER BY added_at DESC, id DESC LIMIT ?',
    args: [guildId, limit],
  });
  return result.rows;
}

// All entries for a single session (= all anime sharing the same watched_date)
async function getEntriesForDate(guildId, watchedDate) {
  await db.ready;
  const result = await db.client.execute({
    sql: 'SELECT * FROM anime_night_entries WHERE guild_id = ? AND watched_date = ?',
    args: [guildId, watchedDate],
  });
  return result.rows;
}

// Replaces the entire title list of a session, optionally moving it to a new date
// in the same operation (delete + reinsert, in a single batch).
async function replaceSession(guildId, oldDate, newDate, titles, editedBy) {
  await db.ready;
  const addedAt = Date.now();

  const statements = [
    { sql: 'DELETE FROM anime_night_entries WHERE guild_id = ? AND watched_date = ?', args: [guildId, oldDate] },
    ...titles.map((title) => ({
      sql: `INSERT INTO anime_night_entries (guild_id, title, watched_date, added_at, added_by)
            VALUES (?, ?, ?, ?, ?)`,
      args: [guildId, title, newDate, addedAt, editedBy],
    })),
  ];

  await db.client.batch(statements, 'write');
}

// Moves a session to a new date without touching its titles. If the new date
// matches an existing session, they naturally merge (grouping is purely by date).
async function updateSessionDate(guildId, oldDate, newDate) {
  await db.ready;
  await db.client.execute({
    sql: 'UPDATE anime_night_entries SET watched_date = ? WHERE guild_id = ? AND watched_date = ?',
    args: [newDate, guildId, oldDate],
  });
}

module.exports = {
  addEntries,
  getAllEntries,
  getLastEntries,
  getEntriesForDate,
  replaceSession,
  updateSessionDate,
};

File: discord-bot-python/src/features/animenight/anime_night_repository.py
import time

from src.database import db


def _now_ms() -> int:
    return int(time.time() * 1000)


async def add_entries(guild_id: str, titles: list[str], watched_date: str, added_by: str) -> None:
    """Inserts multiple anime entries in a single batch (all sharing the same watched
    date and "added at" timestamp, since they come from one /animenight add call)."""
    await db.ready()
    added_at = _now_ms()

    statements = [
        (
            "INSERT INTO anime_night_entries (guild_id, title, watched_date, added_at, added_by) "
            "VALUES (?, ?, ?, ?, ?)",
            [guild_id, title, watched_date, added_at, added_by],
        )
        for title in titles
    ]

    await db.client.batch(statements)


async def get_all_entries(guild_id: str) -> list[dict]:
    await db.ready()
    result = await db.client.execute(
        "SELECT * FROM anime_night_entries WHERE guild_id = ? ORDER BY id ASC", [guild_id]
    )
    return db.rows_as_dicts(result)


async def get_last_entries(guild_id: str, limit: int) -> list[dict]:
    await db.ready()
    result = await db.client.execute(
        "SELECT * FROM anime_night_entries WHERE guild_id = ? ORDER BY added_at DESC, id DESC LIMIT ?",
        [guild_id, limit],
    )
    return db.rows_as_dicts(result)


async def get_entries_for_date(guild_id: str, watched_date: str) -> list[dict]:
    """All entries for a single session (= all anime sharing the same watched_date)."""
    await db.ready()
    result = await db.client.execute(
        "SELECT * FROM anime_night_entries WHERE guild_id = ? AND watched_date = ?", [guild_id, watched_date]
    )
    return db.rows_as_dicts(result)


async def replace_session(guild_id: str, old_date: str, new_date: str, titles: list[str], edited_by: str) -> None:
    """Replaces the entire title list of a session, optionally moving it to a new date
    in the same operation (delete + reinsert, in a single batch)."""
    await db.ready()
    added_at = _now_ms()

    statements = [
        ("DELETE FROM anime_night_entries WHERE guild_id = ? AND watched_date = ?", [guild_id, old_date]),
        *[
            (
                "INSERT INTO anime_night_entries (guild_id, title, watched_date, added_at, added_by) "
                "VALUES (?, ?, ?, ?, ?)",
                [guild_id, title, new_date, added_at, edited_by],
            )
            for title in titles
        ],
    ]

    await db.client.batch(statements)


async def update_session_date(guild_id: str, old_date: str, new_date: str) -> None:
    """Moves a session to a new date without touching its titles. If the new date
    matches an existing session, they naturally merge (grouping is purely by date)."""
    await db.ready()
    await db.client.execute(
        "UPDATE anime_night_entries SET watched_date = ? WHERE guild_id = ? AND watched_date = ?",
        [new_date, guild_id, old_date],
    )
